## Final Project

Lydia Holley

BINF-6112-001

04/06/2023

lholley4@uncc.edu

In [85]:
import pandas as pd
import numpy as np

# Part 1: College Rankings

Examine the following pages and rankings:

https://www.usnews.com/best-colleges

https://www.princetonreview.com/college-rankings/best-colleges

https://www.forbes.com/top-colleges/

Your goal in this part of the assignment is to examine the methodology of these rankings and attempt to recreate it with the scorecard data. Your process should be well justified and presented through the use of markdowns

## describing Part 1

For part one I will emulate the methodology from the Forbes article because:
    the US News methodology has too many discrepancies from our data
    the princeton methodology does not contain a cohesive "best" list in order
    
the forbes method uses the following criteria (adjusted using US News criteria to fit available data):
1. Alumni Salary (20%)
    - MD_EARN_WNE_P10
2. Debt (15%) 
    - GRAD_DEBT_MDN
3. Graduation Rate (15%) 
    - OMAWDP6_FTFT
4. High School Performance (8%)
    - "SATVRMID","SATMTMID","SATWRMID","ACTCMMID"
5. Percent Full Time Faculty (2%)
    - PFTFAC
6. Faculty Salary (10%) 
    - AVGFACSAL
7. Retention Rate (15%) 
    - RET_FT4
8. proportion in debt (10%)
    - FTFTPCTFLOAN
9. Pell Graduation Rate (5%)
    - OMAWDP8_PELL_FTFT

In [214]:
#create function to get scores - user inputs file and the rest is completed for them
def forbesRank(fileName):
    
    #import the file
    colleges = pd.read_csv(fileName)
    
    #create the table with relevent data called "forbes"
    #drop NaN values and set index as the university names
    #make sure all data is float so you can do computational methods on them
    #initialize a new dataFrame called score
    forbes = colleges[["INSTNM","MD_EARN_WNE_P10","GRAD_DEBT_MDN","OMAWDP6_FTFT","SATVRMID","SATMTMID","SATWRMID","ACTCMMID","PFTFAC","AVGFACSAL","RET_FT4","FTFTPCTFLOAN","OMAWDP8_PELL_FTFT"]].replace("PrivacySuppressed",np.nan).dropna().set_index('INSTNM')
    forbes = forbes.astype(float)
    score = forbes
    
    #loop through the columns, find max value of each column
    #create new column for each category with the original value divided by the max value
    for column in forbes:
        maximum = forbes[column].max()
        if column == "MD_EARN_WNE_P10":
            score = score.assign(Alumni_Salary=score[column]/maximum)
        elif column == "GRAD_DEBT_MDN":
            score = score.assign(Debt=score[column]/maximum)
        elif column == "OMAWDP6_FTFT":
            score = score.assign(Grad_Rate=score[column]/maximum)
        elif column == "SATVRMID":
            score = score.assign(SAT_Reading=score[column]/maximum)
        elif column == "SATMTMID":
            score = score.assign(SAT_Math=score[column]/maximum)
        elif column == "SATWRMID":
            score = score.assign(SAT_Writing=score[column]/maximum)
        elif column == "ACTCMMID":
            score = score.assign(ACT_Comp=score[column]/maximum)
        elif column == "PFTFAC":
            score = score.assign(Pct_Fac=score[column]/maximum)
        elif column == "AVGFACSAL":
            score = score.assign(Fac_Salary=score[column]/maximum)
        elif column == "RET_FT4":
            score = score.assign(Ret_Rate=score[column]/maximum)
        elif column == "FTFTPCTFLOAN":
            score = score.assign(Pct_Loan=score[column]/maximum)
        elif column == "OMAWDP8_PELL_FTFT":
            score = score.assign(Pell_Grad_Rate=score[column]/maximum)
    
    #get scores by adding the proportion multiplied the weight for each category
    #the ones related to debt are 1 - debt proportion because debt is undesirable, thus leading to a lower score
    score = score.assign(Score=(score["Alumni_Salary"]*0.20) + ((1 - score["Debt"])*0.15) + (score["Grad_Rate"]*0.15) + (score["SAT_Reading"]*0.02) + (score["SAT_Math"]*0.02) + (score["SAT_Writing"]*0.02) + (score["ACT_Comp"]*0.02) + (score["Pct_Fac"]*0.02) + (score["Fac_Salary"]*0.10) + (score["Ret_Rate"]*0.15) + ((1 - score["Pct_Loan"])*0.10) + (score["Pell_Grad_Rate"]*0.05))
    
    #keep only relevent data
    score = score[["MD_EARN_WNE_P10","GRAD_DEBT_MDN","OMAWDP6_FTFT","SATVRMID","SATMTMID","SATWRMID","ACTCMMID","PFTFAC","AVGFACSAL","RET_FT4","FTFTPCTFLOAN","OMAWDP8_PELL_FTFT","Score"]]
    
    #add rank
    score['rank']=score['Score'].rank(ascending=False)
    rank = score.sort_values(by='rank',ascending=True)
    
    #return dataFrame
    return rank

In [215]:
#check if function worked
forbes_score = forbesRank("Most-Recent-Cohorts-Institution.csv")

/var/folders/n6/gbphvvd12sq8410bcnt0m5pc0000gn/T/ipykernel_26134/612610795.py:5: DtypeWarning: Columns (9,441,442,443,444,445,446,447,448,449,450,451,452,453,454,455,456,457,458,459,460,461,462,463,464,465,466,467,468,469,470,471,472,473,474,475,476,477,478,479,480,481,482,483,484,485,486,487,488,489,490,491,492,493,494,495,496,497,498,499,500,501,502,503,504,505,506,507,508,509,510,511,512,513,514,515,516,517,518,519,520,521,522,523,524,525,526,527,528,529,530,531,532,533,534,535,536,537,538,539,540,541,542,543,544,545,546,547,548,549,550,551,552,553,554,555,556,557,558,559,560,561,562,563,564,565,566,567,568,569,570,571,572,573,574,575,576,577,578,579,580,581,582,583,584,585,586,587,588,589,590,591,592,593,594,595,596,597,598,599,600,601,602,603,604,605,606,607,608,609,610,611,612,613,614,615,616,617,618,619,620,621,622,623,624,625,626,627,628,629,630,631,632,633,634,635,636,637,638,639,640,641,642,643,644,645,646,647,648,649,650,651,652,653,654,655,656,657,658,659,660,661,662,663,66

In [219]:
forbes_score = forbes_score[["MD_EARN_WNE_P10","GRAD_DEBT_MDN","OMAWDP6_FTFT","RET_FT4","FTFTPCTFLOAN","Score","rank"]]
forbes_score[:10]

MD_EARN_WNE_P10  GRAD_DEBT_MDN  \
INSTNM                                                                  
Massachusetts Institute of Technology         111222.0        13418.0   
California Institute of Technology            112166.0        17747.0   
Princeton University                           95689.0        10450.0   
Stanford University                            97798.0        12000.0   
University of Pennsylvania                    103246.0        16763.0   
Harvard University                             84918.0        12665.0   
Duke University                                93115.0        13500.0   
University of Chicago                          76730.0        14601.0   
Rice University                                77683.0        12000.0   
Johns Hopkins University                       83287.0        12750.0   

                                       OMAWDP6_FTFT  RET_FT4  FTFTPCTFLOAN  \
INSTNM                                                                       
Massachusetts Institute of Technology        0.9416   0.9764        0.0735   
California Institute of Technology           0.9202   0.9362        0.0426   
Princeton University                         0.9638   0.8330        0.0210   
Stanford University                          0.9438   0.8575        0.0612   
University of Pennsylvania                   0.9549   0.9499        0.1141   
Harvard University                           0.9777   0.7567        0.0261   
Duke University                              0.9562   0.9688        0.1883   
University of Chicago                        0.9416   0.9965        0.0643   
Rice University                              0.9471   0.9666        0.0916   
Johns Hopkins University                     0.9299   0.9651        0.1464   

                                          Score  rank  
INSTNM                                                 
Massachusetts Institute of Technology  0.885531   1.0  
California Institute of Technology     0.867093   2.0  
Princeton University                   0.861268   3.0  
Stanford University                    0.859069   4.0  
University of Pennsylvania             0.842963   5.0  
Harvard University                     0.831660   6.0  
Duke University                        0.831511   7.0  
University of Chicago                  0.821084   8.0  
Rice University                        0.817893   9.0  
Johns Hopkins University               0.813497  10.0

## Results: Part 1


According to my rendition of their methodology the top 5 schools are:
- 1. Massachusetts Institute of Technology
- 2. California Institute of Technology	
- 3. Princeton University	
- 4. Stanford University
- 5. University of Pennsylvania	
- 6-10. Harvard University, Duke University, University of Chicago, Rice University, Johns Hopkins University		

Top ten from the forbes list is:
- 1. Massachusetts Institute of Technology
- 2. Stanford University
- 2. University of California, Berkely
- 4. Princeton University
- 5. Columbia University
- 6-10. University of California LA, Williams College, Yale University, Duke University, University of Pennsylvania

My rendition of the methodology seems to have been pretty accurate. Many of the same schools on these lists and the top school is the same as well. I would guess the differences come from the missing data from the forbes list (I imagine this is the criteria that ensures the 'correct' schools stay on top). 

# Part 2: New Ranking systems
    
It should not be a surprise to anyone that there are.. limitations to how these rankings are determined. They have a narrow focus that centers around a predetermined result (the top schools must always remain the top schools). In this section, you will determine your own ranking system using the aspects of colleges that are important to you. In a markdown cell, tell me what factors are important to you and why, then rerun the rankings. Display the new results and contrast them with the results of Part 1.

## describing Part 2

The following available data is important to me if i were looking for a college and why:
1. Alumni Salary - 20%
    - MD_EARN_WNE_P10
    - This inidcates how well the school prepares students for life after academia
2. Debt - 20%
    - GRAD_DEBT_MDN
    - this indicates how affordable to school is, the impact of their financial aid, and the ability of students to afford the level of debt given to the students
3. Graduation Rate - 15%
    - OMAWDP6_FTFT
    - this is important because graduating is the goal
4. Retention Rate - 15%
    - RET_FT4
    - retention rate is super important because it means that students like the school enough to return the following years
5. Under ED Monitoring - 10%
    - HCM2
    - it is a bad sign if this is true, so this is important as well. 
    - I will consider being under ED as a 0 
    - not being under monitoring as a 1 
        - will have to do (1 - HCM2) for this
6. Proportion in Debt - 10%
    - FTFTPCTFLOAN
    - this is important because the lower this is the more likely students are to be getting aid besides just loans
7. Faculty Salary - 10%
    - AVGFACSAL
    - this is important because well payed faculty will care more about their jobs

In [208]:
#create function to get scores - user inputs file and the rest is completed for them
def myRank(fileName):
    
    #import the file
    colleges = pd.read_csv(fileName)
    
    #create the table with relevent data called "forbes"
    #drop NaN values and set index as the university names
    #make sure all data is float so you can do computational methods on them
    #initialize a new dataFrame called score
    rank = colleges[["INSTNM","MD_EARN_WNE_P10","GRAD_DEBT_MDN","OMAWDP6_FTFT","RET_FT4","HCM2","FTFTPCTFLOAN","AVGFACSAL"]].replace("PrivacySuppressed",np.nan).dropna().set_index('INSTNM')
    rank = rank.astype(float)
    score = rank
    
    #loop through the columns, find max value of each column
    #create new column for each category with the original value divided by the max value
    for column in rank:
        maximum = rank[column].max()
        if column == "MD_EARN_WNE_P10":
            score = score.assign(Alumni_Salary=score[column]/maximum)
        elif column == "GRAD_DEBT_MDN":
            score = score.assign(Debt=score[column]/maximum)
        elif column == "OMAWDP6_FTFT":
            score = score.assign(Grad_Rate=score[column]/maximum)
        elif column == "RET_FT4":
            score = score.assign(Ret_Rate=score[column]/maximum)
        elif column == "HCM2":
            score = score.assign(ED_Mon=score[column])
        elif column == "FTFTPCTFLOAN":
            score = score.assign(Pct_Loan=score[column]/maximum)
        elif column == "AVGFACSAL":
            score = score.assign(Fac_Sal=score[column]/maximum)
    
    #get scores by adding the proportion multiplied the weight for each category
    #the ones related to debt are 1 - debt proportion because debt is undesirable, thus leading to a lower score
    score = score.assign(Score=(score["Alumni_Salary"]*0.20) + ((1 - score["Debt"])*0.20) + (score["Grad_Rate"]*0.15) + (score["Ret_Rate"]*0.15) + ((1 - score["ED_Mon"])*0.10) + ((1 - score["Pct_Loan"])*0.10) + (score["Fac_Sal"]*0.10))
    
    #keep only relevent data
    score = score[["MD_EARN_WNE_P10","GRAD_DEBT_MDN","OMAWDP6_FTFT","RET_FT4","HCM2","FTFTPCTFLOAN","AVGFACSAL","Score"]]
    
    #add rank
    score['rank']=score['Score'].rank(ascending=False)
    rank = score.sort_values(by='rank',ascending=True)
    
    #return dataFrame
    return rank

In [209]:
#check if function worked
my_score = myRank("Most-Recent-Cohorts-Institution.csv")

/var/folders/n6/gbphvvd12sq8410bcnt0m5pc0000gn/T/ipykernel_26134/876995197.py:5: DtypeWarning: Columns (9,441,442,443,444,445,446,447,448,449,450,451,452,453,454,455,456,457,458,459,460,461,462,463,464,465,466,467,468,469,470,471,472,473,474,475,476,477,478,479,480,481,482,483,484,485,486,487,488,489,490,491,492,493,494,495,496,497,498,499,500,501,502,503,504,505,506,507,508,509,510,511,512,513,514,515,516,517,518,519,520,521,522,523,524,525,526,527,528,529,530,531,532,533,534,535,536,537,538,539,540,541,542,543,544,545,546,547,548,549,550,551,552,553,554,555,556,557,558,559,560,561,562,563,564,565,566,567,568,569,570,571,572,573,574,575,576,577,578,579,580,581,582,583,584,585,586,587,588,589,590,591,592,593,594,595,596,597,598,599,600,601,602,603,604,605,606,607,608,609,610,611,612,613,614,615,616,617,618,619,620,621,622,623,624,625,626,627,628,629,630,631,632,633,634,635,636,637,638,639,640,641,642,643,644,645,646,647,648,649,650,651,652,653,654,655,656,657,658,659,660,661,662,663,66

In [213]:
my_score[:10]

MD_EARN_WNE_P10  GRAD_DEBT_MDN  \
INSTNM                                                                   
Massachusetts Institute of Technology          111222.0        13418.0   
Princeton University                            95689.0        10450.0   
California Institute of Technology             112166.0        17747.0   
Stanford University                             97798.0        12000.0   
University of Pennsylvania                     103246.0        16763.0   
Duke University                                 93115.0        13500.0   
Harvard University                              84918.0        12665.0   
University of Chicago                           76730.0        14601.0   
Rice University                                 77683.0        12000.0   
Franklin W Olin College of Engineering         132969.0        15846.0   

                                        OMAWDP6_FTFT  RET_FT4  HCM2  \
INSTNM                                                                
Massachusetts Institute of Technology         0.9416   0.9764   0.0   
Princeton University                          0.9638   0.8330   0.0   
California Institute of Technology            0.9202   0.9362   0.0   
Stanford University                           0.9438   0.8575   0.0   
University of Pennsylvania                    0.9549   0.9499   0.0   
Duke University                               0.9562   0.9688   0.0   
Harvard University                            0.9777   0.7567   0.0   
University of Chicago                         0.9416   0.9965   0.0   
Rice University                               0.9471   0.9666   0.0   
Franklin W Olin College of Engineering        0.9259   0.8118   0.0   

                                        FTFTPCTFLOAN  AVGFACSAL     Score  \
INSTNM                                                                      
Massachusetts Institute of Technology         0.0735    19640.0  0.884795   
Princeton University                          0.0210    20835.0  0.866482   
California Institute of Technology            0.0426    20228.0  0.864864   
Stanford University                           0.0612    20971.0  0.860514   
University of Pennsylvania                    0.1141    18702.0  0.848427   
Duke University                               0.1883    16663.0  0.832709   
Harvard University                            0.0261    21143.0  0.832667   
University of Chicago                         0.0643    18432.0  0.826223   
Rice University                               0.0916    16378.0  0.822355   
Franklin W Olin College of Engineering        0.3294    12021.0  0.818749   

                                        rank  
INSTNM                                        
Massachusetts Institute of Technology    1.0  
Princeton University                     2.0  
California Institute of Technology       3.0  
Stanford University                      4.0  
University of Pennsylvania               5.0  
Duke University                          6.0  
Harvard University                       7.0  
University of Chicago                    8.0  
Rice University                          9.0  
Franklin W Olin College of Engineering  10.0

## Results: Part 2

According to my rendition of the forbes methodology the top 10 schools are:
- 1. Massachusetts Institute of Technology
- 2. California Institute of Technology	
- 3. Princeton University	
- 4. Stanford University
- 5. University of Pennsylvania	
- 6-10. Harvard University, Duke University, University of Chicago, Rice University, Johns Hopkins University		

Top ten from the forbes list is:
- 1. Massachusetts Institute of Technology
- 2. Stanford University
- 2. University of California, Berkely
- 4. Princeton University
- 5. Columbia University 
- 6-10. University of California LA, Williams College, Yale University, Duke University, University of Pennsylvania

#### Top 10 from my own methodology:
- 1. Massachusetts Institute of Technology
- 2. Princeton University
- 3. California Institute of Technology
- 4. Stanford University
- 5. University of Pennsylvania
- 6-10. Duke University, Harvard University, University of Chicago, Rice University, Franklin W Olin College of Engineering	

My methodology seems to was closer to the rendition I did of forbes than the actual forbes list as well. Many of the same schools on these lists and the top school is the same as well. I would guess the differences come from the missing data from the forbes list (I imagine this is the criteria that ensures the 'correct' schools stay on top) for the forbes ranking. The difference between my rendition of forbes and my own methodology was the inclusion of ED Monitoring and the exclusion of some data like high school rankings. 

# Part 3: Program rankings
    
Using your two ranking systems, apply that methodology (as closely as you can) to individual programs of study. You can choose Bioinformatics programs if you want, but if you want to pick something else that's also fine. Display and contrast the results.

## describing part 3

the forbes method uses the following criteria (with our equivalent listed below it):
1. Alumni Salary (20%)
    - MD_EARN_WNE_P10
        - Field of Study Median Earnings
        - EARN_MDN_4YR
2. Debt (15%) 
    - GRAD_DEBT_MDN
        - Field of Study Median Total Debt for All Federal Loans
        - DEBT_ALL_STGP_ANY_MDN
3. Graduation Rate (15%) 
    - OMAWDP6_FTFT
        - Field of Study Number of Graduates
        - IPEDSCOUNT2
4. High School Performance (8%)
    - "SATVRMID","SATMTMID","SATWRMID","ACTCMMID"
        - no equivalent
5. Percent Full Time Faculty (2%)
    - PFTFAC
        - no equivalent
6. Faculty Salary (10%) 
    - AVGFACSAL
        - no equivalent
7. Retention Rate (15%) 
    - RET_FT4
        - no equivalent
8. proportion in debt (10%)
    - FTFTPCTFLOAN
        - BBRR4_FED_COMP_N
9. Pell Graduation Rate (5%)
    - OMAWDP8_PELL_FTFT  
        - no equivalent

#### for part 3 I will use the following criteria and weights (only one bc not enough data to differentiate a forbes and personal ranking):
1. Alumni Salary (30%)
    - MD_EARN_WNE_P10
        - Field of Study Median Earnings
        - EARN_MDN_4YR
2. Debt (30%) 
    - GRAD_DEBT_MDN
        - Field of Study Median Total Debt for All Federal Loans
        - DEBT_ALL_STGP_ANY_MDN
3. Graduation Rate (20%) 
    - OMAWDP6_FTFT
        - Field of Study Number of Graduates
        - IPEDSCOUNT2
8. proportion in debt (20%)
    - FTFTPCTFLOAN
        - BBRR4_FED_COMP_N

In [206]:
#create function to get scores - user inputs file and the rest is completed for them
def fosRank(fileName):
    
    #import the file
    colleges = pd.read_csv(fileName)
    
    #create the table with relevent data called "fos"
    #drop NaN values 
    #make sure all data is float so you can do computational methods on them
    #initialize a new dataFrame called score
    fos = colleges[["INSTNM","CIPCODE","CIPDESC","EARN_MDN_4YR","DEBT_ALL_STGP_ANY_MDN","IPEDSCOUNT2","BBRR4_FED_COMP_N"]].replace("PrivacySuppressed",np.nan).dropna()
    score = fos
    
    #loop through the columns, find max value of each column
    #create new column for each category with the original value divided by the max value
    for column in fos:
        if column == "EARN_MDN_4YR":
            maximum = fos[column].astype(float).max()
            score = score.assign(Alumni_Salary=score[column].astype(float)/maximum)
        elif column == "DEBT_ALL_STGP_ANY_MDN":
            maximum = fos[column].astype(float).max()
            score = score.assign(Debt=score[column].astype(float)/maximum)
        elif column == "IPEDSCOUNT2":
            maximum = fos[column].astype(float).max()
            score = score.assign(Grad_Rate=score[column].astype(float)/maximum)
        elif column == "BBRR4_FED_COMP_N":
            maximum = fos[column].astype(float).max()
            score = score.assign(Prop_Debt=score[column].astype(float)/maximum)
        else:
            continue
    
    #get scores by adding the proportion multiplied the weight for each category
    #the ones related to debt are 1 - debt proportion because debt is undesirable, thus leading to a lower score
    score = score.assign(Score=(score["Alumni_Salary"]*0.30) + ((1 - score["Debt"])*0.30) + (score["Grad_Rate"]*0.20) + ((1 - score["Prop_Debt"])*0.20))
    
    #keep only relevent data
    score = score[["INSTNM","CIPCODE","CIPDESC","EARN_MDN_4YR","DEBT_ALL_STGP_ANY_MDN","IPEDSCOUNT2","BBRR4_FED_COMP_N","Score"]]
    
    #add rank
    score['rank']=score['Score'].rank(ascending=False)
    rank = score.sort_values(by='rank',ascending=True)

    #return dataFrame
    return rank

In [229]:
fos = fosRank("Most-Recent-Cohorts-Field-of-Study.csv")

In [236]:
comp_sci = fos[fos["CIPDESC"] == "Computer Science."]
comp_sci['cs_rank']=comp_sci['rank'].rank(ascending=True)
comp_sci[:10]

/var/folders/n6/gbphvvd12sq8410bcnt0m5pc0000gn/T/ipykernel_26134/1192220449.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  comp_sci['cs_rank']=comp_sci['rank'].rank(ascending=True)


INSTNM  CIPCODE            CIPDESC  \
158949                       Brown University     1107  Computer Science.   
198934                    Stanford University     1107  Computer Science.   
149017             Carnegie Mellon University     1107  Computer Science.   
79519   Massachusetts Institute of Technology     1107  Computer Science.   
12039       University of California-Berkeley     1107  Computer Science.   
113069                     Cornell University     1107  Computer Science.   
126071                        Duke University     1107  Computer Science.   
167643                  Vanderbilt University     1107  Computer Science.   
155114             University of Pennsylvania     1107  Computer Science.   
17454                     Harvey Mudd College     1107  Computer Science.   

       EARN_MDN_4YR DEBT_ALL_STGP_ANY_MDN  IPEDSCOUNT2 BBRR4_FED_COMP_N  \
158949       218525                 11500        229.0               46   
198934       200950                 10399        307.0               40   
149017       247552                 21442        237.0               84   
79519        199774                 12000        424.0               68   
12039        178867                 13900        845.0              131   
113069       185679                 14698        396.0              149   
126071       159845                 13500        339.0               83   
167643       164139                 14500        147.0               21   
155114       146438                 13500        107.0               17   
17454        183524                 22949         47.0               46   

           Score  rank  cs_rank  
158949  0.708939   2.0      1.0  
198934  0.695224   3.0      2.0  
149017  0.691655   5.0      3.0  
79519   0.687458   7.0      4.0  
12039   0.660210  11.0      5.0  
113069  0.654051  12.0      6.0  
126071  0.629069  15.0      7.0  
167643  0.626076  17.0      8.0  
155114  0.609042  25.0      9.0  
17454   0.602781  30.0     10.0

In [234]:
comp_sci[comp_sci["INSTNM"] == "University of North Carolina at Charlotte"]

INSTNM  CIPCODE            CIPDESC  \
128723  University of North Carolina at Charlotte     1107  Computer Science.   

       EARN_MDN_4YR DEBT_ALL_STGP_ANY_MDN  IPEDSCOUNT2 BBRR4_FED_COMP_N  \
128723        82583                 23250        539.0              282   

           Score    rank  cs_rank  
128723  0.484963  4730.0    142.0

## Results: Part 3

Top 10 Computer Science programs from this own methodology:
- 1. Brown University
- 2. Stanford University
- 3. Carnegie Mellon University
- 4. Massachusetts Institute of Technology
- 5. University of California-Berkeley
- 6-10. Cornell University, Duke University, Vanderbilt University, University of Pennsylvania, Harvey Mudd College	

UNCC rank for the Computer Science programs:
- 142. University of North Carolina at Charlotte

I was going to do bioinformatics, but there is no data on that program, so I chose Computer Science. The top ten Comp Sci schools have some overlap between the top ten from the other methodologies/ranks. UNCC is actually in the bottom half of the chosen criteria for this program (142 out of 247). This appears to be because the salary of graudates is much lower and the number of students in debt/amount of debt is much higher.

# Part 4: Summarizing results
    
Finally, pick an institution that matters to you in some way. UNC Charlotte or where you did your undergrad are good choices, but you don't necessarily have to choose those. Where did that school fall in your ranking systems? If it was low in the rankings, what factors caused it to be? What steps could that institution take to improve its rankings? You can use past data as well to show how that institution is improving or failing to improve.

## part 4 description

I was going to choose George Mason University because that is where I went to undergrad, but it must have some NaN datatypes and was removed from the data frames. Instead I will choose UNCC for this. 

In [220]:
forbes_score.loc[['University of North Carolina at Charlotte']]

MD_EARN_WNE_P10  GRAD_DEBT_MDN  \
INSTNM                                                                      
University of North Carolina at Charlotte          52252.0        21668.0   

                                           OMAWDP6_FTFT  RET_FT4  \
INSTNM                                                             
University of North Carolina at Charlotte        0.5842   0.8472   

                                           FTFTPCTFLOAN     Score   rank  
INSTNM                                                                    
University of North Carolina at Charlotte        0.5099  0.553174  206.0

In [221]:
my_score.loc[['University of North Carolina at Charlotte']]

MD_EARN_WNE_P10  GRAD_DEBT_MDN  \
INSTNM                                                                      
University of North Carolina at Charlotte          52252.0        21668.0   

                                           OMAWDP6_FTFT  RET_FT4  HCM2  \
INSTNM                                                                   
University of North Carolina at Charlotte        0.5842   0.8472   0.0   

                                           FTFTPCTFLOAN  AVGFACSAL     Score  \
INSTNM                                                                         
University of North Carolina at Charlotte        0.5099     9886.0  0.599065   

                                            rank  
INSTNM                                            
University of North Carolina at Charlotte  419.0

In [222]:
my_score

MD_EARN_WNE_P10  GRAD_DEBT_MDN  \
INSTNM                                                                  
Massachusetts Institute of Technology         111222.0        13418.0   
Princeton University                           95689.0        10450.0   
California Institute of Technology            112166.0        17747.0   
Stanford University                            97798.0        12000.0   
University of Pennsylvania                    103246.0        16763.0   
...                                                ...            ...   
Bacone College                                 36274.0        29375.0   
Arkansas Baptist College                       26710.0        26375.0   
South University-Austin                        36710.0        25379.0   
The Art Institute of Atlanta                   34110.0        30750.0   
Martin University                              27347.0        41604.0   

                                       OMAWDP6_FTFT  RET_FT4  HCM2  \
INSTNM                                                               
Massachusetts Institute of Technology        0.9416   0.9764   0.0   
Princeton University                         0.9638   0.8330   0.0   
California Institute of Technology           0.9202   0.9362   0.0   
Stanford University                          0.9438   0.8575   0.0   
University of Pennsylvania                   0.9549   0.9499   0.0   
...                                             ...      ...   ...   
Bacone College                               0.1026   0.4237   1.0   
Arkansas Baptist College                     0.1679   0.4286   1.0   
South University-Austin                      0.1250   0.2857   1.0   
The Art Institute of Atlanta                 0.1409   0.4545   1.0   
Martin University                            0.0417   0.2000   0.0   

                                       FTFTPCTFLOAN  AVGFACSAL     Score  \
INSTNM                                                                     
Massachusetts Institute of Technology        0.0735    19640.0  0.884795   
Princeton University                         0.0210    20835.0  0.866482   
California Institute of Technology           0.0426    20228.0  0.864864   
Stanford University                          0.0612    20971.0  0.860514   
University of Pennsylvania                   0.1141    18702.0  0.848427   
...                                             ...        ...       ...   
Bacone College                               0.5342     4560.0  0.279633   
Arkansas Baptist College                     0.6517     4303.0  0.275274   
South University-Austin                      0.7778     6365.0  0.263725   
The Art Institute of Atlanta                 0.7973     5353.0  0.258472   
Martin University                            0.9000     4950.0  0.237983   

                                         rank  
INSTNM                                         
Massachusetts Institute of Technology     1.0  
Princeton University                      2.0  
California Institute of Technology        3.0  
Stanford University                       4.0  
University of Pennsylvania                5.0  
...                                       ...  
Bacone College                         1806.0  
Arkansas Baptist College               1807.0  
South University-Austin                1808.0  
The Art Institute of Atlanta           1809.0  
Martin University                      1810.0  

[1810 rows x 9 columns]

## Results: Part 4

UNCC rank from the forbes list is:
- 252. University of North Carolina, Charlotte

According to my rendition of the forbes methodology the UNCC rank is:
- 206. University of North Carolina at Charlotte	

UNCC rank from my own methodology:
- 419. University of North Carolina at Charlotte	

UNCC did the best on my rendition of the forbes methodology and did the worst with my own methodology. The places where UNCC could improve are graduation rate, faculty salary, and improve the financial aid packages so fewer students require loans and accumulate less debt. The graduation rate is very low and the amount of debt is very high.  

In [ ]:
table.shape()
(row,column)
